// react-localize-redux/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _Translate = require('./Translate');

Object.defineProperty(exports, 'Translate', {
  enumerable: true,
  get: function get() {
    return _Translate.Translate;
  }
});

var _withLocalize = require('./withLocalize');

Object.defineProperty(exports, 'withLocalize', {
  enumerable: true,
  get: function get() {
    return _withLocalize.withLocalize;
  }
});

var _LocalizeProvider = require('./LocalizeProvider');

Object.defineProperty(exports, 'LocalizeProvider', {
  enumerable: true,
  get: function get() {
    return _LocalizeProvider.LocalizeProvider;
  }
});

var _LocalizeContext = require('./LocalizeContext');

Object.defineProperty(exports, 'LocalizeContext', {
  enumerable: true,
  get: function get() {
    return _LocalizeContext.LocalizeContext;
  }
});

var _localize = require('./localize');

Object.defineProperty(exports, 'localizeReducer', {
  enumerable: true,
  get: function get() {
    return _localize.localizeReducer;
  }
});
Object.defineProperty(exports, 'initialize', {
  enumerable: true,
  get: function get() {
    return _localize.initialize;
  }
});
Object.defineProperty(exports, 'addTranslation', {
  enumerable: true,
  get: function get() {
    return _localize.addTranslation;
  }
});
Object.defineProperty(exports, 'addTranslationForLanguage', {
  enumerable: true,
  get: function get() {
    return _localize.addTranslationForLanguage;
  }
});
Object.defineProperty(exports, 'setActiveLanguage', {
  enumerable: true,
  get: function get() {
    return _localize.setActiveLanguage;
  }
});
Object.defineProperty(exports, 'getTranslate', {
  enumerable: true,
  get: function get() {
    return _localize.getTranslate;
  }
});
Object.defineProperty(exports, 'getActiveLanguage', {
  enumerable: true,
  get: function get() {
    return _localize.getActiveLanguage;
  }
});
Object.defineProperty(exports, 'getLanguages', {
  enumerable: true,
  get: function get() {
    return _localize.getLanguages;
  }
});
Object.defineProperty(exports, 'getTranslations', {
  enumerable: true,
  get: function get() {
    return _localize.getTranslations;
  }
});
Object.defineProperty(exports, 'getOptions', {
  enumerable: true,
  get: function get() {
    return _localize.getOptions;
  }
});
